src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1505,12 +1636,179 @@ namespace Microsoft.FSharp.Core
                     override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
 
+            let canUseDefaultEqualityComparer (ty:Type) =
+                let processed = System.Collections.Generic.HashSet ()
+
+                let rec recurse idx (types:array<Type>) =
+                    if idx = types.Length then true
+                    else
+                        let ty = get types idx
+                        if not (processed.Add ty) then
+                            recurse (idx+1) types
+                        else
+                            let isValidGenericType ifNotType fullname =
+                                if not (ty.IsGenericType && ty.GetGenericTypeDefinition().FullName.Equals fullname)
+                                then ifNotType
+                                else recurse 0 (ty.GetGenericArguments ())
+                    
+                            // avoid any types that need special handling in GenericEqualityObj
+                            true 
+                            && ty.IsSealed  // covers enum and value types
+                            && not ty.IsArray
+                            && not (ty.Equals typeof<String>)
+                            && not (ty.Equals typeof<Decimal>)
+                            && not (ty.Equals typeof<float>)
+                            && not (ty.Equals typeof<float32>)
+                            && isValidGenericType true "System.Nullable`1"
+                            && not (typeof<IStructuralEquatable>.IsAssignableFrom ty
+                                    && not (false
+                                            || isValidGenericType false "System.ValueTuple`1"
+                                            || isValidGenericType false "System.ValueTuple`2"
+                                            || isValidGenericType false "System.ValueTuple`3"
+                                            || isValidGenericType false "System.ValueTuple`4"
+                                            || isValidGenericType false "System.ValueTuple`5"
+                                            || isValidGenericType false "System.ValueTuple`6"
+                                            || isValidGenericType false "System.ValueTuple`7"
+                                            || isValidGenericType false "System.ValueTuple`8"))
+                            && recurse (idx+1) types
+
+                recurse 0 [|ty|]
+
+            let tryGetFSharpEqualityComparer (er:bool) (ty:Type) : obj =
+                match er, ty with
+                | _, ty when ty.Equals typeof<string> -> box {
+                    new EqualityComparer<string>() with
+                        member __.Equals (x,y) = System.String.Equals (x, y)
+                        member __.GetHashCode x = HashString x }
+
+                | _, ty when ty.Equals typeof<decimal> -> box {
```

Comment:
Nit: This is odd formatting for an object expression, we would always use

    { new X with 
        member ...
    }

or

    { new X with 
        member ...     }

If piping is defined at this point then the box can be like this:

    { new X with 
        member ...     }
    |> box

Reply:
I don't think piping is available in prim_types but I'll see... I'll move the brace down anyway...